Table of Contents

1. Import Libraries

2. Import Data

3. Exercise 4.5 Data Consistency Checks (on df_prods dataframe)

4. Task 4.5 Data Consistency Checks (on df_ords dataframe) Step 2. Run df.describe function

5. Step 3. Check for mixed data types

6. Step 5. Check for missing values

7. Step 6. Address missing values appropriately

8. Step 7. Check for duplicates in orders dataframe

9. Step 8. Address duplicates if found

10. Export dataframes

1. Import Libraries

In [103]:
# Import Libraries
import pandas as pd
import numpy as np
import os

2. Import data

In [104]:
# Create pathname
path = r'/Users/ianfleming/Desktop/05-2025 Instacart Basket Analysis'

In [105]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [106]:
# Check import success
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [107]:
# Import Orders dataframe
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [108]:
# Check import success
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

3. Exercise 4.5 Data Consistency Checks

In [109]:
# Create a dataframe
df_test = pd.DataFrame()

In [110]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [111]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [112]:
# Check for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [113]:
# Create a subset
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [114]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [115]:
# Number of rows in df_prods
df_prods.shape

(49693, 5)

In [116]:
#Run explanatory analysis on the df_products dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [117]:
#Find the median of the df_prods dataframe
df_prods['prices'].median()

7.1

In [118]:
#Find outliers in the df_prods dataframe.
df_prods[df_prods['prices'] > df_prods['prices'].mean() + (2 * df_prods['prices'].std())]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [119]:
# Replace outlier values in the prices column with the meadian value of 7.1.
df_prods = df_prods.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [120]:
# Check the df_prods dataframe after imputation
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682245
std    14343.717401     38.316774       5.850282      4.200159
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

During my exploratory analysis of the df_prods dataframe, I noticed unusually high values in the prices column. The mean price was 9.99, but the standard deviation was over 450, which is extremely large compared to the mean. This indicated the likely presence of outliers.

To confirm this, I identified values that were more than 2 standard deviations above the mean, a common statistical method for detecting outliers. This revealed two extreme values: 14,900 and 99,999 — far outside the normal price range for products.

Since these values would heavily skew averages and distort analysis, I decided to treat them as outliers. I replaced them with the median price, which was 7.1. The median is less sensitive to extreme values than the mean, making it a more reliable measure of central tendency in this case.

After replacing these outliers, the mean price decreased to 7.6, reflecting a more accurate average for the data. 

In [121]:
# create a new clean prods dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]


In [122]:
df_prods_clean.shape

(49677, 5)

In [123]:
# Find duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [124]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [125]:
# Check number of rows in clean products dataframe

In [126]:
df_prods_clean.shape

(49677, 5)

In [127]:
# Create dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [128]:
df_prods_clean_no_dups.shape

(49672, 5)

In [129]:
# Export cleaned products dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index = False)

Task 4.5

Step 2. Run df.describe function


4. Step 2. Run df.describe function

In [130]:
# Import Orders dataframe
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [131]:
# Check import success
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [132]:
# Check stats of orders dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

(A) The 'Unnamed:0' column could be dropped as it is the dataframe's index and matches the 'order_id' column. (B) in the 'days_since_prior_order' column the count is 3.214874 while all other columns are the same at 3.421083. So there are 206,209 missing values that need to be handled.  

Step 3. Check for mixed data types

5. Step 3. Check for mixed data types

In [133]:
# Check for mixed data types in df_ords dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

/var/folders/8s/_430jnv556z0gjsmqhjpf76m0000gn/T/ipykernel_4870/1121569957.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/8s/_430jnv556z0gjsmqhjpf76m0000gn/T/ipykernel_4870/1121569957.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/8s/_430jnv556z0gjsmqhjpf76m0000gn/T/ipykernel_4870/1121569957.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/8s/_430jnv556z0gjsmqhjpf76m0000gn/T/ipykernel_4870/1121569957.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0]

In [134]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


Couldn't get an output with the above code so tried others to get a further understanding.  

In [135]:
for col in df_ords.columns:
    types_found = df_ords[col].map(type).value_counts()
    if len(types_found) > 1:
        print(f"\nColumn '{col}' has mixed types:")
        print(types_found)
        

In [136]:
for col in df_ords.columns:
    types_found = df_ords[col].map(type).value_counts()
    print(f"\nColumn '{col}' types:")
    print(types_found)


Column 'Unnamed: 0' types:
Unnamed: 0
<class 'int'>    3421083
Name: count, dtype: int64

Column 'order_id' types:
order_id
<class 'int'>    3421083
Name: count, dtype: int64

Column 'user_id' types:
user_id
<class 'int'>    3421083
Name: count, dtype: int64

Column 'order_number' types:
order_number
<class 'int'>    3421083
Name: count, dtype: int64

Column 'orders_day_of_week' types:
orders_day_of_week
<class 'int'>    3421083
Name: count, dtype: int64

Column 'order_hour_of_day' types:
order_hour_of_day
<class 'int'>    3421083
Name: count, dtype: int64

Column 'days_since_prior_order' types:
days_since_prior_order
<class 'float'>    3421083
Name: count, dtype: int64


Step 4. Fix mixed datatype if found

I didn't find any

Step 5. Check for missing values

6. Step 5. Check for missing values

In [137]:
# check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 missing values in the days_since_prior_order column. The missing values in the days_since_prior_order column likely correspond to customers' first-ever orders. Since there is no previous order to compare to, it cannot compute a time difference. Therefore, the missing values rows should not be deleted.

Step 6. Address missing values appropriately

7. Step 6. Address missing values appropriately

In [138]:
# Create a subset of the dataframe with only the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [139]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [140]:
# Because the missing values correspond with customer's first orders (can see that 'order_number' is 1, so check all rows allign with the order number == 1
df_ords_nan[df_ords_nan['days_since_prior_order'].isna()]['order_number'].value_counts()


order_number
1    206209
Name: count, dtype: int64

Yes they all order_numbers correspond to == 1 so that means that the missing values are customer's first orders. Inform the client about the missing values and the important reason to leave them in the dataframe.

In [141]:
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [142]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 8 columns]

Step 7. Check for duplicates in orders dataframe

8. Step 7. Check for duplicates in orders dataframe

In [143]:
# Check for duplicates in orders dataframe
df_ords_dups = df_ords[df_ords.duplicated()]

There are no duplicates in the dataframe

Step 8. Address duplicates if found

9. Step 8. Address duplicates if found

In [144]:
# Check to see if anything else can be cleaned
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 8 columns]

In [145]:
# Address & Drop 'Unnamed: 0' column
df_ords_clean = df_ords.drop('Unnamed: 0', axis = 1)


In [146]:
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 7 columns]

'df_ords_clean' contains no duplicate rows or mixed datatypes, no unnecessary columns and the missing values are addressed, reported and remain in the dataframe.

10. Export dataframes

In [147]:
# Export cleaned dataframe for orders to the prepared data folder
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [148]:
# Export cleaned dataframe for orders to the prepared data folder (as the above saved the index column)
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index=False)

